### PR TITLE
[Validator] Fixed `Traverse` constraint reference

### LINKED ISSUE #12967: Fix the documentation of the Traverse constraint
https://github.com/symfony/symfony-docs/pull/11310#issuecomment-520484292

The current code from https://symfony.com/doc/4.4/reference/constraints/Traverse.html:

<details>

```php
<?php
// src/Entity/Book.php
namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * @ORM\Entity
 * @Assert\Traverse
 */
class Book
{
    /**
     * @var Author
     *
     * @ORM\ManyToOne(targetEntity="App\Entity\Author")
     */
    protected $author;

    /**
     * @var Editor
     *
     * @ORM\ManyToOne(targetEntity="App\Entity\Editor")
     */
    protected $editor;

    // ...
}
```

</details>

throws a `\Symfony\Component\Validator\Exception\ConstraintDefinitionException` with message `Traversal was enabled for "App\Entity\Book", but this class does not implement "\Traversable".` (from `vendor/symfony/validator/Validator/RecursiveContextualValidator.php`).

From https://github.com/symfony/symfony/pull/10287:

<details>

> ##### Control traversal at class level (symfony/symfony#8617)
> 
> Currently, it is possible whether to traverse a `Traversable` object or not in the `Valid` constraint:
> 
> ```
> /**
>  * @Assert\Valid(traverse=true)
>  */
> private $tags = new TagList();
> ```
> 
> (actually, `true` is the default)
> 
> In this way, the validator will iterate the `TagList` instance and validate each of the contained objects. You can also set "traverse" to `false` to disable iteration.
> 
> What if you want to specify, that `TagList` instances should always (or never) be traversed? That's currently not possible.
> 
> With this PR, you can do the following:
> 
> ```
> /**
>  * @Assert\Traverse(false)
>  */
> class TagList implements \IteratorAggregate
> {
>     // ...
> }
> ```

</details>

I understand that the `Traverse` constraint is not at all a "shorthand" for `Valid` constraints on all nested objects, but rather a means to always disable validator traversal on a custom `\Traversable` class. (But I may be wrong)

### PR DESCRIPTION
**Pull request by HeahDude**
I'm sorry I've messed up things in #11310. Now I see that I missed #12967, this fixes it.

### CODE REVIEW

File: reference/constraints/Traverse.rst [hunk 1]
```diff
@@ -1,10 +1,10 @@
 Traverse
 ========
 
-Objects do not validate nested objects by default unless explicitly using
-this constraint.
-If only specific nested objects should be validated by cascade, consider
-using the :doc:`/reference/constraints/Valid` instead.
+Objects properties are only validated if they are accessible, either by being
+public or having public accessor methods (e.g a public getter).
```
Comment (guilliamxavier) [inline reference/constraints/Traverse.rst:5]:
```suggestion
public or having public accessor methods (e.g. a public getter).
```

File: reference/constraints/Traverse.rst [hunk 2]
```diff
@@ -17,17 +17,14 @@ using the :doc:`/reference/constraints/Valid` instead.
...
 
 .. configuration-block::
 
     .. code-block:: php-annotations
 
-        // src/AppBundle/Entity/Book.php
+        // src/AppBundle/Entity/BookCollection.php
         namespace AppBundle\Entity;
 
         use Doctrine\ORM\Mapping as ORM;
```
Comment (guilliamxavier) [inline reference/constraints/Traverse.rst:30]:
```suggestion
        use Doctrine\Common\Collections\ArrayCollection;
        use Doctrine\Common\Collections\Collection;
        use Doctrine\ORM\Mapping as ORM;
```

File: reference/constraints/Traverse.rst [hunk 3]
```diff
@@ -37,31 +34,66 @@ configure the ``Traverse`` constraint on the ``Book`` class.
...
              *
-             * @ORM\ManyToOne(targetEntity="AppBundle\Entity\Author")
+             * @Assert\NotBlank
              */
-            protected $author;
+            protected $name = '';
 
             /**
-             * @var Editor
+             * @var Book[]
```
Comment (guilliamxavier) [inline reference/constraints/Traverse.rst:49]:
```suggestion
             * @var Collection|Book[]
```

File: reference/constraints/Traverse.rst [hunk 4]
```diff
@@ -37,31 +34,66 @@ configure the ``Traverse`` constraint on the ``Book`` class.
...
+            {
+                foreach ($this->books as $book) {
+                    if ($book->isAuthoredBy($author)) {
+                        yield $book;
+                    }
+                }
+            }
+
+            // neither the method above nor any other specific getter
+            // could be used to validated all nested books
```
Comment (guilliamxavier) [inline reference/constraints/Traverse.rst:83]:
```suggestion
            // could be used to validated all nested books;
```

File: reference/constraints/Traverse.rst [hunk 5]
```diff
@@ -84,15 +116,23 @@ configure the ``Traverse`` constraint on the ``Book`` class.
         use Symfony\Component\Validator\Constraints as Assert;
         use Symfony\Component\Validator\Mapping\ClassMetadata;
 
-        class Book
+        class BookCollection
```
Comment (guilliamxavier) [inline reference/constraints/Traverse.rst:119]: I can't comment on collapsed lines (https://github.com/isaacs/github/issues/1655), I wanted to suggest on old line 81:
```diff
-        // src/AppBundle/Entity/Book.php
+        // src/AppBundle/Entity/BookCollection.php
```

File: reference/constraints/Traverse.rst [hunk 6]
```diff
@@ -37,31 +34,66 @@ configure the ``Traverse`` constraint on the ``Book`` class.
...
+            // ... setter for name, adder and remover for books
+
+            // the name can be validated by calling the getter
+            public function getName(): string
+            {
+                return $this->name;
+            }
+
+            /**
+             * @return Book[]|\Generator The books for a given author
```
Comment (guilliamxavier) [inline reference/constraints/Traverse.rst:71]:
```suggestion
             * @return \Generator|Book[] The books for a given author
```
(real type first)
(actually it ought to be `\Generator<Book>` if we had generics...)

**Review by guilliamxavier — COMMENTED** [review]
Thank you @HeahDude. I can't really say for the use case (hopefully @xabbuh will) but I have left a few editorial comments.

File: reference/constraints/Traverse.rst (hunk 5 above)
Comment (HeahDude) [reply reference/constraints/Traverse.rst:119]: Good catch! Thanks

**Comment by HeahDude** [comment]
Thanks @guilliamxavier for the review, I've addressed your comment and also added more doc about the default and the `traverse` option.

File: reference/constraints/Traverse.rst [hunk 7]
```diff
@@ -1,10 +1,10 @@
 Traverse
 ========
 
-Objects do not validate nested objects by default unless explicitly using
-this constraint.
-If only specific nested objects should be validated by cascade, consider
-using the :doc:`/reference/constraints/Valid` instead.
+Objects properties are only validated if they are accessible, either by being
```
Comment (xabbuh) [inline reference/constraints/Traverse.rst:4]: Object properties [...]

File: reference/constraints/Traverse.rst [hunk 8]
```diff
@@ -71,28 +105,63 @@ configure the ``Traverse`` constraint on the ``Book`` class.
...
+The ``groups`` option is not available for this constraint.
+
+
+``traverse``
+~~~~~~~~~~~~
+
+**type**: ``bool`` **default**: ``true``
+
+Instances of `\Traversable` are traversed by default, use this option to
+disable validating::
```
Comment (xabbuh) [inline reference/constraints/Traverse.rst:152]: What about using a `configuration-block` to also document this option for the other formats?

File: reference/constraints/Traverse.rst (hunk 7 above)
Comment (HeahDude) [reply reference/constraints/Traverse.rst]: Fixed

File: reference/constraints/Traverse.rst (hunk 8 above)
Comment (HeahDude) [reply reference/constraints/Traverse.rst]: Good call, done!

File: reference/constraints/Traverse.rst [hunk 9]
```diff
@@ -71,28 +105,105 @@ configure the ``Traverse`` constraint on the ``Book`` class.
...
 
+The ``groups`` option is not available for this constraint.
+
+
+``traverse``
+~~~~~~~~~~~~
+
+**type**: ``bool`` **default**: ``true``
+
+Instances of `\Traversable` are traversed by default, use this option to
```
Comment (guilliamxavier) [inline reference/constraints/Traverse.rst:151]: DOCtor-RST suggests to double the backslash but I would rather double the backticks

**Review by xabbuh — APPROVED** [review]

**Pull request merged**